README.md
# Timepiece  [![CocoaPods Compatible](https://img.shields.io/cocoapods/v/Timepiece.svg)](https://cocoapods.org/pods/Timepiece) [![Build Status](http://img.shields.io/travis/naoty/Timepiece/master.svg?style=flat)](https://travis-ci.org/naoty/Timepiece)
Intuitive date handling in Swift

## Features
* :bulb: **Intuitive**: Timepiece provides a set of helpers to make date handling easier.
* :clock9: **Correct**: Using Foundation API correctly, Timepiece helps to calculate dates correctly without deep understanding.
* :package: **Small**: Timepiece has only 4 file. You can read the inside of this easily.

## Requirements

* Swift 3.0+
* iOS 8.0+ / macOS 10.9+ / tvOS 9.0+ / watchOS 2.0+

## Usage

### Initialization

```swift
Date(era: 235, year: 26, month: 8, day: 14, hour: 20, minute: 25, second: 43, nanosecond: 0, on: Calendar(identifier: .japanese))
Date(year: 2014, month: 8, day: 14, hour: 20, minute: 25, second: 43, nanosecond: 0)
Date(year: 2014, month: 8, day: 14, hour: 20, minute: 25, second: 43)
Date(year: 2014, month: 8, day: 14)

Date.today()
Date.yesterday()
Date.tomorrow()
```

### Calculation

```swift
now + 1.year
now - 2.months
now + (3.weeks - 4.days + 5.hours)

1.year.later
1.year.ago
```

### Change

```swift
now.changed(year: 2014)
now.changed(weekday: 1)
now.truncated([.minute, .second, .nanosecond])
now.truncated(from: .day)
```

### Formating

```swift
now.stringIn(dateStyle: .long, timeStyle: .medium)
now.dateString(in: .medium)
now.timeString(in: .short)
now.stringIn(dateFormat: "'LDML representation:' yyyy/MM/dd"))

3.days.string(in: .full)
```

### Parsing

```swift
"2014/8/14".date(inFormat: "yyyy/MM/dd")
"2014-08-14T20:25:43+0900".dateInISO8601Format()
```

## Installation

### CocoaPods

```ruby
# Podfile
pod "Timepiece"
```

### Carthage

```ruby
# Cartfile
github "naoty/Timepiece"
```

## Contribution

1. Fork
2. Create your feature branch (git checkout -b my-new-feature)
3. Commit your changes (git commit -am 'Add some feature')
4. Push to the branch (git push origin my-new-feature)
5. Create new Pull Request

## License

Timepiece is available under the MIT license. See the LICENSE file for more info.

## Author

[naoty](https://github.com/naoty)



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## Unreleased

## 1.2.1
Released on 2017-08-21.

### Added
* `stringIn(dateFormat:)`, which creates a `String` representation of `Date` instance based on LDML format string.

## 1.2.0
Released on 2017-04-30.

### Added
* Support Swift Package Manager. [#83](https://github.com/naoty/Timepiece/pull/83)
* `truncated(_:)`, which creates a `Date` instance by truncating the components. [#82](https://github.com/naoty/Timepiece/pull/82)
* `truncated(from:)`, which creates a `Date` instance by truncating the components from a given component. [#82](https://github.com/naoty/Timepiece/pull/82)

### Fixed
* Rename `string(inDateStyle:andTimeStyle:)` to `stringIn(dateStyle:timeStyle:)`.
* Rename `dateInISO8601Format(withOptions:)` to `dateInISO8601Format(with:)`.

## 1.1.0
Released on 2017-01-06.

### Added
* `changed(year:month:day:hour:minute:second:nanosecond:)`, which creates a `Date` instance by changing receiver's date components. [#77](https://github.com/naoty/Timepiece/pull/77)
* `changed(weekday:)`, which creates a `Date` instance by changing receiver's weekday. [#77](https://github.com/naoty/Timepiece/pull/77)

## 1.0.2
Released on 2016-12-20.

### Fixed
* Fix testDateInISO8601Format() availability. [#74](https://github.com/naoty/Timepiece/pull/74).
* Specify Swift version for the compilation of watchOS target. [#79](https://github.com/naoty/Timepiece/pull/79).

## 1.0.1
Released on 2016-11-05.

### Changed
* Create instance of date components instead of mutable variables. [#67](https://github.com/naoty/Timepiece/pull/67).

### Fixed
* Fix App Store submission issue. [#71](https://github.com/naoty/Timepiece/pull/71).

## 1.0.0
Released on 2016-10-30.

### Added
* New initializers for `Date`.
* `+` and `-` between `DateComponents`
* Methods to format `Date` with date style.
* A method to format `DateComponents` with units style.
* A method to parse `String` using `ISO8601DateFormatter`.

### Removed
* `Date.date()` constructor methods.
* `Duration`.
* Time zone support for each `Date`.
* Properties to get beginning and end of periods such as `beginningOfYear`, `endOfMonth`.
* `change()`.

### Changed
* For `Int`, `year`, `month` and so on creates new `DateComponents` instance instead of `Duration`.


Tests/TimepieceTests/Date+TimepieceTests.swift
//
//  Date+TimepieceTests.swift
//  Timepiece
//
//  Created by Naoto Kaneko on 10/15/16.
//  Copyright © 2016 Naoto Kaneko. All rights reserved.
//

import Timepiece
import XCTest

class DateTests: XCTestCase {
    private var sample: Date {
        return Date(year: 2014, month: 8, day: 15, hour: 20, minute: 25, second: 43)
    }
    
    func testYear() {
        XCTAssertEqual(sample.year, 2014)
    }

    func testMonth() {
        XCTAssertEqual(sample.month, 8)
    }

    func testDay() {
        XCTAssertEqual(sample.day, 15)
    }

    func testHour() {
        XCTAssertEqual(sample.hour, 20)
    }

    func testMinute() {
        XCTAssertEqual(sample.minute, 25)
    }

    func testSecond() {
        XCTAssertEqual(sample.second, 43)
    }

    func testNanosecond() {
        XCTAssertEqual(sample.nanosecond, 0)
    }

    func testWeekday() {
        XCTAssertEqual(sample.weekday, 6)
    }

    func testInit() {
        let heisei = 235
        var japaneseCalendar = Calendar(identifier: .japanese)
        japaneseCalendar.timeZone = TimeZone(identifier: "Asia/Tokyo")!
        let date = Date(era: heisei, year: 26, month: 8, day: 15, hour: 20, minute: 25, second: 43, nanosecond: 0, on: japaneseCalendar)

        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "GMT")!
        let dateComponents = calendar.dateComponents([.year, .month, .day, .hour, .minute, .second, .nanosecond, .weekday], from: date)
        XCTAssertEqual(dateComponents.year, 2014)
        XCTAssertEqual(dateComponents.month, 8)
        XCTAssertEqual(dateComponents.day, 15)
        XCTAssertEqual(dateComponents.hour, 11)
        XCTAssertEqual(dateComponents.minute, 25)
        XCTAssertEqual(dateComponents.second, 43)
        XCTAssertEqual(dateComponents.nanosecond, 0)
        XCTAssertEqual(dateComponents.weekday, 6)
    }

    func testInitForTime() {
        let date = Date(year: 2014, month: 8, day: 15, hour: 20, minute: 25, second: 43)

        let dateComponents = Calendar.current.dateComponents([.year, .month, .day, .hour, .minute, .second, .nanosecond, .weekday], from: date)
        XCTAssertEqual(dateComponents.year, 2014)
        XCTAssertEqual(dateComponents.month, 8)
        XCTAssertEqual(dateComponents.day, 15)
        XCTAssertEqual(dateComponents.hour, 20)
        XCTAssertEqual(dateComponents.minute, 25)
        XCTAssertEqual(dateComponents.second, 43)
        XCTAssertEqual(dateComponents.nanosecond, 0)
        XCTAssertEqual(dateComponents.weekday, 6)
    }

    func testInitForDate() {
        let date = Date(year: 2014, month: 8, day: 15)

        let dateComponents = Calendar.current.dateComponents([.year, .month, .day, .hour, .minute, .second, .nanosecond, .weekday], from: date)
        XCTAssertEqual(dateComponents.year, 2014)
        XCTAssertEqual(dateComponents.month, 8)
        XCTAssertEqual(dateComponents.day, 15)
        XCTAssertEqual(dateComponents.weekday, 6)
    }

    func testToday() {
        let today = Date.today()
        let now = Date()

        XCTAssertEqual(today.year, now.year)
        XCTAssertEqual(today.month, now.month)
        XCTAssertEqual(today.day, now.day)
        XCTAssertEqual(today.hour, 0)
        XCTAssertEqual(today.minute, 0)
        XCTAssertEqual(today.second, 0)
    }

    func testYesterday() {
        let yesterday = Date.yesterday()
        let date = Date() - 1.day

        XCTAssertEqual(yesterday.year, date?.year)
        XCTAssertEqual(yesterday.month, date?.month)
        XCTAssertEqual(yesterday.day, date?.day)
        XCTAssertEqual(yesterday.hour, 0)
        XCTAssertEqual(yesterday.minute, 0)
        XCTAssertEqual(yesterday.second, 0)
    }

    func testTomorrow() {
        let tomorrow = Date.tomorrow()
        let date = Date() + 1.day

        XCTAssertEqual(tomorrow.year, date?.year)
        XCTAssertEqual(tomorrow.month, date?.month)
        XCTAssertEqual(tomorrow.day, date?.day)
        XCTAssertEqual(tomorrow.hour, 0)
        XCTAssertEqual(tomorrow.minute, 0)
        XCTAssertEqual(tomorrow.second, 0)
    }

    func testPlusDateComponents() {
        let date = Date(year: 2016, month: 1, day: 1)

        var dateComponents = DateComponents()
        dateComponents.month = 1

        let added = date + dateComponents
        XCTAssertEqual(added?.month, 2)
    }

    func testMinusDateComponents() {
        let date = Date(year: 2016, month: 1, day: 1)

        var dateComponents = DateComponents()
        dateComponents.month = 1

        let added = date - dateComponents
        XCTAssertEqual(added?.month, 12)
    }

    func testChanged() {
        let date = Date(year: 2014, month: 8, day: 14, hour: 20, minute: 25, second: 43, nanosecond: 0)
        let newDate = date.changed(year: 2017)

        XCTAssertEqual(newDate?.year, 2017)
        XCTAssertEqual(newDate?.month, 8)
        XCTAssertEqual(newDate?.day, 14)
        XCTAssertEqual(newDate?.hour, 20)
        XCTAssertEqual(newDate?.minute, 25)
        XCTAssertEqual(newDate?.second, 43)
        XCTAssertEqual(newDate?.nanosecond, 0)
    }

    func testChangedByWeekday() {
        let date = Date(year: 2014, month: 8, day: 14, hour: 20, minute: 25, second: 43, nanosecond: 0)
        let newDate = date.changed(weekday: 7)

        XCTAssertEqual(newDate?.year, 2014)
        XCTAssertEqual(newDate?.month, 8)
        XCTAssertEqual(newDate?.day, 16)
        XCTAssertEqual(newDate?.hour, 20)
        XCTAssertEqual(newDate?.minute, 25)
        XCTAssertEqual(newDate?.second, 43)
        XCTAssertEqual(newDate?.nanosecond, 0)
    }

    func testTruncated() {
        let date = Date(year: 2014, month: 8, day: 14, hour: 20, minute: 25, second: 43, nanosecond: 0)
        let newDate = date.truncated([.hour, .minute, .second, .nanosecond])

        XCTAssertEqual(newDate?.year, 2014)
        XCTAssertEqual(newDate?.month, 8)
        XCTAssertEqual(newDate?.day, 14)
        XCTAssertEqual(newDate?.hour, 0)
        XCTAssertEqual(newDate?.minute, 0)
        XCTAssertEqual(newDate?.second, 0)
        XCTAssertEqual(newDate?.nanosecond, 0)
    }

    func testTruncatedFrom() {
        let date = Date(year: 2014, month: 8, day: 14, hour: 20, minute: 25, second: 43, nanosecond: 0)
        let newDate = date.truncated(from: .month)

        XCTAssertEqual(newDate?.year, 2014)
        XCTAssertEqual(newDate?.month, 1)
        XCTAssertEqual(newDate?.day, 1)
        XCTAssertEqual(newDate?.hour, 0)
        XCTAssertEqual(newDate?.minute, 0)
        XCTAssertEqual(newDate?.second, 0)
        XCTAssertEqual(newDate?.nanosecond, 0)
    }

    func testStringInStyles() {
        let sampleString = sample.stringIn(dateStyle: .short, timeStyle: .short)
        XCTAssertEqual(sampleString, "8/15/14, 8:25 PM")
    }

    func testDateString() {
        let sampleString = sample.dateString(in: .short)
        XCTAssertEqual(sampleString, "8/15/14")
    }

    func testTimeString() {
        let sampleString = sample.timeString(in: .short)
        XCTAssertEqual(sampleString, "8:25 PM")
    }
    
    func testStringInFormat() {
        let sampleString = sample.stringIn(dateFormat: "yyyy-MM-dd")
        XCTAssertEqual(sampleString, "2014-08-15")
    }
}


Sources/Date+Timepiece.swift
//
//  Date+Timepiece.swift
//  Timepiece
//
//  Created by Naoto Kaneko on 10/2/16.
//  Copyright © 2016 Naoto Kaneko. All rights reserved.
//

import Foundation

extension Date {
    /// The year.
    public var year: Int {
        return dateComponents.year!
    }

    /// The month.
    public var month: Int {
        return dateComponents.month!
    }

    /// The day.
    public var day: Int {
        return dateComponents.day!
    }

    /// The hour.
    public var hour: Int {
        return dateComponents.hour!
    }

    /// The minute.
    public var minute: Int {
        return dateComponents.minute!
    }

    /// The second.
    public var second: Int {
        return dateComponents.second!
    }

    /// The nanosecond.
    public var nanosecond: Int {
        return dateComponents.nanosecond!
    }

    /// The weekday.
    public var weekday: Int {
        return dateComponents.weekday!
    }

    private var dateComponents: DateComponents {
        return calendar.dateComponents([.era, .year, .month, .day, .hour, .minute, .second, .nanosecond, .weekday], from: self)
    }

    // Returns user's calendar to be used to return `DateComponents` of the receiver.
    private var calendar: Calendar {
        return .current
    }

    /// Creates a new instance with specified date components.
    ///
    /// - parameter era:        The era.
    /// - parameter year:       The year.
    /// - parameter month:      The month.
    /// - parameter day:        The day.
    /// - parameter hour:       The hour.
    /// - parameter minute:     The minute.
    /// - parameter second:     The second.
    /// - parameter nanosecond: The nanosecond.
    /// - parameter calendar:   The calendar used to create a new instance.
    ///
    /// - returns: The created `Date` instance.
    public init(era: Int?, year: Int, month: Int, day: Int, hour: Int, minute: Int, second: Int, nanosecond: Int, on calendar: Calendar) {
        let now = Date()
        var dateComponents = calendar.dateComponents([.era, .year, .month, .day, .hour, .minute, .second, .nanosecond], from: now)
        dateComponents.era = era
        dateComponents.year = year
        dateComponents.month = month
        dateComponents.day = day
        dateComponents.hour = hour
        dateComponents.minute = minute
        dateComponents.second = second
        dateComponents.nanosecond = nanosecond

        let date = calendar.date(from: dateComponents)
        self.init(timeInterval: 0, since: date!)
    }

    /// Creates a new instance with specified date componentns.
    ///
    /// - parameter year:       The year.
    /// - parameter month:      The month.
    /// - parameter day:        The day.
    /// - parameter hour:       The hour.
    /// - parameter minute:     The minute.
    /// - parameter second:     The second.
    /// - parameter nanosecond: The nanosecond. `0` by default.
    ///
    /// - returns: The created `Date` instance.
    public init(year: Int, month: Int, day: Int, hour: Int, minute: Int, second: Int, nanosecond: Int = 0) {
        self.init(era: nil, year: year, month: month, day: day, hour: hour, minute: minute, second: second, nanosecond: nanosecond, on: .current)
    }

    /// Creates a new Instance with specified date components
    ///
    /// - parameter year:  The year.
    /// - parameter month: The month.
    /// - parameter day:   The day.
    ///
    /// - returns: The created `Date` instance.
    public init(year: Int, month: Int, day: Int) {
        self.init(year: year, month: month, day: day, hour: 0, minute: 0, second: 0)
    }

    /// Creates a new instance representing today.
    ///
    /// - returns: The created `Date` instance representing today.
    public static func today() -> Date {
        let now = Date()
        return Date(year: now.year, month: now.month, day: now.day)
    }

    /// Creates a new instance representing yesterday
    ///
    /// - returns: The created `Date` instance representing yesterday.
    public static func yesterday() -> Date {
        return (today() - 1.day)!
    }

    /// Creates a new instance representing tomorrow
    ///
    /// - returns: The created `Date` instance representing tomorrow.
    public static func tomorrow() -> Date {
        return (today() + 1.day)!
    }

    /// Creates a new instance added a `DateComponents`
    ///
    /// - parameter left:  The date.
    /// - parameter right: The date components.
    ///
    /// - returns: The created `Date` instance.
    public static func + (left: Date, right: DateComponents) -> Date? {
        return Calendar.current.date(byAdding: right, to: left)
    }

    /// Creates a new instance subtracted a `DateComponents`
    ///
    /// - parameter left:  The date.
    /// - parameter right: The date components.
    ///
    /// - returns: The created `Date` instance.
    public static func - (left: Date, right: DateComponents) -> Date? {
        return Calendar.current.date(byAdding: -right, to: left)
    }

    /// Creates a new instance by changing the date components
    ///
    /// - Parameters:
    ///   - year: The year.
    ///   - month: The month.
    ///   - day: The day.
    ///   - hour: The hour.
    ///   - minute: The minute.
    ///   - second: The second.
    ///   - nanosecond: The nanosecond.
    /// - Returns: The created `Date` instnace.
    public func changed(year: Int? = nil, month: Int? = nil, day: Int? = nil, hour: Int? = nil, minute: Int? = nil, second: Int? = nil, nanosecond: Int? = nil) -> Date? {
        var dateComponents = self.dateComponents
        dateComponents.year = year ?? self.year
        dateComponents.month = month ?? self.month
        dateComponents.day = day ?? self.day
        dateComponents.hour = hour ?? self.hour
        dateComponents.minute = minute ?? self.minute
        dateComponents.second = second ?? self.second
        dateComponents.nanosecond = nanosecond ?? self.nanosecond

        return calendar.date(from: dateComponents)
    }

    /// Creates a new instance by changing the year.
    ///
    /// - Parameter year: The year.
    /// - Returns: The created `Date` instance.
    public func changed(year: Int) -> Date? {
        return changed(year: year, month: nil, day: nil, hour: nil, minute: nil, second: nil, nanosecond: nil)
    }

    /// Creates a new instance by changing the month.
    ///
    /// - Parameter month: The month.
    /// - Returns: The created `Date` instance.
    public func changed(month: Int) -> Date? {
        return changed(year: nil, month: month, day: nil, hour: nil, minute: nil, second: nil, nanosecond: nil)
    }

    /// Creates a new instance by changing the day.
    ///
    /// - Parameter day: The day.
    /// - Returns: The created `Date` instance.
    public func changed(day: Int) -> Date? {
        return changed(year: nil, month: nil, day: day, hour: nil, minute: nil, second: nil, nanosecond: nil)
    }

    /// Creates a new instance by changing the hour.
    ///
    /// - Parameter hour: The hour.
    /// - Returns: The created `Date` instance.
    public func changed(hour: Int) -> Date? {
        return changed(year: nil, month: nil, day: nil, hour: hour, minute: nil, second: nil, nanosecond: nil)
    }

    /// Creates a new instance by changing the minute.
    ///
    /// - Parameter minute: The minute.
    /// - Returns: The created `Date` instance.
    public func changed(minute: Int) -> Date? {
        return changed(year: nil, month: nil, day: nil, hour: nil, minute: minute, second: nil, nanosecond: nil)
    }

    /// Creates a new instance by changing the second.
    ///
    /// - Parameter second: The second.
    /// - Returns: The created `Date` instance.
    public func changed(second: Int) -> Date? {
        return changed(year: nil, month: nil, day: nil, hour: nil, minute: nil, second: second, nanosecond: nil)
    }

    /// Creates a new instance by changing the nanosecond.
    ///
    /// - Parameter nanosecond: The nanosecond.
    /// - Returns: The created `Date` instance.
    public func changed(nanosecond: Int) -> Date? {
        return changed(year: nil, month: nil, day: nil, hour: nil, minute: nil, second: nil, nanosecond: nanosecond)
    }

    /// Creates a new instance by changing the weekday.
    ///
    /// - Parameter weekday: The weekday.
    /// - Returns: The created `Date` instance.
    public func changed(weekday: Int) -> Date? {
        return self - (self.weekday - weekday).days
    }

    /// Creates a new instance by truncating the components
    ///
    /// - Parameter components: The components to be truncated.
    /// - Returns: The created `Date` instance.
    public func truncated(_ components: [Calendar.Component]) -> Date? {
        var dateComponents = self.dateComponents

        for component in components {
            switch component {
            case .month:
                dateComponents.month = 1
            case .day:
                dateComponents.day = 1
            case .hour:
                dateComponents.hour = 0
            case .minute:
                dateComponents.minute = 0
            case .second:
                dateComponents.second = 0
            case .nanosecond:
                dateComponents.nanosecond = 0
            default:
                continue
            }
        }
        
        return calendar.date(from: dateComponents)
    }

    /// Creates a new instance by truncating the components
    ///
    /// - Parameter component: The component to be truncated from.
    /// - Returns: The created `Date` instance.
    public func truncated(from component: Calendar.Component) -> Date? {
        switch component {
        case .month:
            return truncated([.month, .day, .hour, .minute, .second, .nanosecond])
        case .day:
            return truncated([.day, .hour, .minute, .second, .nanosecond])
        case .hour:
            return truncated([.hour, .minute, .second, .nanosecond])
        case .minute:
            return truncated([.minute, .second, .nanosecond])
        case .second:
            return truncated([.second, .nanosecond])
        case .nanosecond:
            return truncated([.nanosecond])
        default:
            return self
        }
    }

    /// Creates a new `String` instance representing the receiver formatted in given date style and time style.
    ///
    /// - parameter dateStyle: The date style.
    /// - parameter timeStyle: The time style.
    ///
    /// - returns: The created `String` instance.
    public func stringIn(dateStyle: DateFormatter.Style, timeStyle: DateFormatter.Style) -> String {
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = dateStyle
        dateFormatter.timeStyle = timeStyle

        return dateFormatter.string(from: self)
    }

    @available(*, unavailable, renamed: "stringIn(dateStyle:timeStyle:)")
    public func string(inDateStyle dateStyle: DateFormatter.Style, andTimeStyle timeStyle: DateFormatter.Style) -> String {
        return stringIn(dateStyle: dateStyle, timeStyle: timeStyle)
    }

    /// Creates a new `String` instance representing the date of the receiver formatted in given date style.
    ///
    /// - parameter dateStyle: The date style.
    ///
    /// - returns: The created `String` instance.
    public func dateString(in dateStyle: DateFormatter.Style) -> String {
        return stringIn(dateStyle: dateStyle, timeStyle: .none)
    }

    /// Creates a new `String` instance representing the time of the receiver formatted in given time style.
    ///
    /// - parameter timeStyle: The time style.
    ///
    /// - returns: The created `String` instance.
    public func timeString(in timeStyle: DateFormatter.Style) -> String {
        return stringIn(dateStyle: .none, timeStyle: timeStyle)
    }
    
    /// Creates a new `String` instance representing the receiver formatted in given custom LDML format string.
    ///
    /// - parameter dateFormat: The LDML format string (see http://www.unicode.org/reports/tr35/tr35-45/tr35-dates.html#Date_Field_Symbol_Table for reference)
    ///
    /// - returns: The created `String` instance.
    public func stringIn(dateFormat: String) -> String {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = dateFormat
        
        return dateFormatter.string(from: self)
    }
}


Timepiece.playground/Contents.swift
import Foundation
import Timepiece

let now = Date()

// Initializer
Date(era: 235, year: 26, month: 8, day: 14, hour: 20, minute: 25, second: 43, nanosecond: 0, on: Calendar(identifier: .japanese))
Date(year: 2014, month: 8, day: 14, hour: 20, minute: 25, second: 43, nanosecond: 0)
Date(year: 2014, month: 8, day: 14, hour: 20, minute: 25, second: 43)
Date(year: 2014, month: 8, day: 14)

Date.today()
Date.yesterday()
Date.tomorrow()

// The properties of Date
now.year
now.month
now.weekday
now.day
now.hour
now.minute
now.second
now.nanosecond

// Calculation
now + 1.year
now - 2.months
now + (3.weeks - 4.days + 5.hours)

1.year.later
1.year.ago

// Change
now.changed(year: 2014)
now.changed(weekday: 1)
now.truncated([.minute, .second, .nanosecond])
now.truncated(from: .day)

// Format
now.stringIn(dateStyle: .long, timeStyle: .medium)
now.dateString(in: .medium)
now.timeString(in: .short)
now.stringIn(dateFormat: "yyyy-MM-dd")

3.days.string(in: .full)

// Parse
"2014/8/14".date(inFormat: "yyyy/MM/dd")
"2014-08-14T20:25:43+0900".dateInISO8601Format()
